File: LL/src/Core/Application/UseCases/CharacterActions/Dtos/Responses/CraftingDtos/TemperingSessionDto.cs
using Application.Common.Mappings;
using AutoMapper;
using Domain.Models.CharacterActions.Sessions;

namespace Application.UseCases.CharacterActions.Dtos.Responses.CraftingDtos;
public class TemperingSessionDto : IMapFrom<TemperingSession>
{
    public DateTimeOffset From { get; set; }
    public DateTimeOffset To { get; set; }
    public TemperingSummary TemperingSummary { get; set; } = null!;
    public IReadOnlyList<TemperingOutcomeEntryDto> Outcomes { get; set; } = [];
    public void Mapping(Profile profile)
    {
        profile.CreateMap<TemperingOutcomeEntry, TemperingOutcomeEntryDto>();
        profile.CreateMap<TemperingSession, TemperingSessionDto>();
    }
}

public sealed class TemperingOutcomeEntryDto
{
    public Guid Id { get; set; }
    public Guid QueueItemId { get; set; }
    public Guid EquipmentInstanceId { get; set; }
    public string EquipmentName { get; set; } = string.Empty;
    public DateTimeOffset OccurredAt { get; set; }
    public Domain.Models.Professions.Crafting.TemperingOutcome Outcome { get; set; }
    public int PotentialSpent { get; set; }
    public int PreviousPotential { get; set; }
    public int NewPotential { get; set; }
    public int PreviousItemXp { get; set; }
    public int NewItemXp { get; set; }
    public bool BecameMasterpiece { get; set; }
    public bool BecameLevelingItem { get; set; }
    public Domain.Models.Items.Rarity PreviousRarity { get; set; }
    public Domain.Models.Items.Rarity NewRarity { get; set; }
    public bool RarityUpgraded { get; set; }
    public bool QualityIncreased { get; set; }
    public Domain.Models.Items.ItemQuality? PreviousQuality { get; set; }
    public Domain.Models.Items.ItemQuality? NewQuality { get; set; }
    public Domain.Models.Attributes.AttributeType? ImprovedStat { get; set; }
    public float? PreviousStatValue { get; set; }
    public float? NewStatValue { get; set; }
}

File: LL/src/Presentation/ll/src/app/shared/models/Dtos/temperingSessionDto.ts
import { AttributeType } from '../enums/attributeType';
import { ItemQuality } from '../enums/itemQuality';
import { Rarity } from '../enums/rarity';

export interface TemperingSessionDto {
  from: Date;
  to: Date;

  temperingSummary: TemperingSummary;
  outcomes?: TemperingOutcomeEntry[];
}

export type TemperingOutcome = 'Critical' | 'Positive' | 'Neutral' | 'Negative';

export interface TemperingOutcomeEntry {
  id: string;
  queueItemId: string;
  equipmentInstanceId: string;
  equipmentName: string;
  occurredAt: string;
  outcome: TemperingOutcome;
  potentialSpent: number;
  previousPotential: number;
  newPotential: number;
  previousItemXp: number;
  newItemXp: number;
  becameMasterpiece: boolean;
  becameLevelingItem: boolean;
  previousRarity: Rarity;
  newRarity: Rarity;
  rarityUpgraded: boolean;
  qualityIncreased: boolean;
  previousQuality?: ItemQuality | null;
  newQuality?: ItemQuality | null;
  improvedStat?: AttributeType | null;
  previousStatValue?: number | null;
  newStatValue?: number | null;
}

export interface TemperingSummary {
  totalItemsCrafted: number;
  masterpieces: number;
  levelingItems: number;
  cursedOutcomes: number;
  qualityIncreases: number;
  totalActions: number;
  totalSoulstones: number;
  craftingExperience: number;
  totalExperience: number;
}
